### PR TITLE
3.7 Allow empty by Flags Validation

### PR DESCRIPTION
**Pull request by ravage84**
I've taken @chinpei215 's work from his PR #12486.

@chinpei215 proposed his changes as a workaround for a vulnerability. But since we couldn't merge the changes into 3.6 or even earlier versions, he withdrew his PR. https://github.com/cakephp/cakephp/pull/12486#issuecomment-419923784

Because I think the changes as they stand are a nice addition, I wanted to re-propose them and open a discussion about them. Not as a workaround for a vulnerability, but as an API improvement for CakePHP 3.7 or 4.0.

I haven't changed anything yet.

### CODE REVIEW

**Review by markstory — APPROVED** [review]
I still think these are good additions that allow us to make validators more expressive and restrictive. @lorenzo @ADmad ?

**Comment by ADmad** [comment]
LGTM, just 2 things:

1. Not too thrilled about the method name `allowEmptyByFlags()`.

2. If we are going to mark `allowEmpty()` as (soft) deprecated then bake should also be updated to use these new methods.

**Comment by ravage84** [comment]
@admad I agree about the naming. Now about ``allowEmotyFor()`` or ``allowEmptyForType()``?

**Comment by ADmad** [comment]
`allowEmptyFor()` works for me.

**Comment by ADmad** [comment]
I took care of the method rename.

**Comment by ravage84** [comment]
@admad thanks and greetings from my army duty.

File: src/Validation/Validator.php [hunk 1]
```diff
@@ -2051,22 +2313,61 @@ protected function _canBeEmpty($field, $context)
...
      *
-     * @param mixed $data value to check against
+     * @param mixed $data Value to check against.
      * @return bool
+     * @deprecated Use isEmpty() instead
      */
     protected function _fieldIsEmpty($data)
     {
-        if (empty($data) && !is_bool($data) && !is_numeric($data)) {
+        return $this->isEmpty($data, self::EMPTY_ALL);
```
Comment (dereuromark) [inline src/Validation/Validator.php]: Can we please use static in these cases?
Especially ALL can at some point be needed as LTB to add another missing one into it.

**Comment by dereuromark** [comment]
"Add `@SInCE` annotations" :trollface: 

Apart from the static missing part this looks good to me.

**Comment by ADmad** [comment]
It was hoping putting it in back ticks would avoid the auto linking but didn't :stuck_out_tongue:.

**Comment by dereuromark** [comment]
I never does for commits... A bit bad for those that have annotation usernames :P But otherwise meeh.

**Comment by ADmad** [comment]
Meh, not my problem. It's the user's / github's fault :slightly_smiling_face:.

**Comment by dereuromark** [comment]
I still think `self::EMPTY_ALL` should better be LateStaticBinding "static::".

**Pull request merged**